### PR TITLE
GIT-253: Report an unchanged settings save as a no-op instead of a failure

### LINKED ISSUE #253: Settings pages report "Sorry, Your changes are not saved!" when the submitted values are unchanged
All nine schema tabs on **Rich Snippets → Dashboard** show an error when you click **Update** without editing a field, even though nothing has failed and the settings are intact.

## Steps to reproduce

1. Go to `wp-admin/admin.php?page=rich_snippet_dashboard`.
2. Open any schema panel (Item Review, Event, Product, …).
3. Click **Update** without changing anything.
4. The page shows **"Sorry, Your changes are not saved!"**

Editing a field to a genuinely new value and saving shows the success notice as expected. Saving again immediately reproduces the error.

## Cause

`update_option()` returns `false` in two different situations: the write failed, **and** the stored value is already identical. The save handlers treat any falsy return as failure.

```php
$status = update_option( 'bsf_review', $args );
display_status( $status );
```

`display_status()` prints the error branch for any falsy `$status`, so an unchanged save is reported as a failure.

Confirmed on a local install:

```
update_option with a CHANGED value  -> bool(true)
update_option with the SAME value   -> bool(false)
```

Simulating an unchanged save for every schema option:

```
bsf_review     -> false  => "Sorry, Your changes are not saved!"
bsf_event      -> false  => "Sorry, Your changes are not saved!"
bsf_person     -> false  => "Sorry, Your changes are not saved!"
bsf_product    -> false  => "Sorry, Your changes are not saved!"
bsf_recipe     -> false  => "Sorry, Your changes are not saved!"
bsf_software   -> false  => "Sorry, Your changes are not saved!"
bsf_video      -> false  => "Sorry, Your changes are not saved!"
bsf_article    -> false  => "Sorry, Your changes are not saved!"
bsf_service    -> false  => "Sorry, Your changes are not saved!"
```

## Affected code

- `admin/index.php` — nine handlers (`bsf_review`, `bsf_event`, `bsf_person`, `bsf_product`, `bsf_recipe`, `bsf_software`, `bsf_video`, `bsf_article`, `bsf_service`) plus the WooCommerce setting, all using `$status = update_option( … ); display_status( $status );`
- `index.php` (`submit_color` AJAX) — returns **"Error occured. Settings were not saved !"** when the same colours are re-saved.

## Proposed fix

Verify the stored result instead of inferring success from the return value:

```php
function bsf_save_snippet_option( $option, $args ) {
	update_option( $option, $args );
	return get_option( $option ) === $args;
}
```

## Additional finding

`event_performer` is listed in the Event save handler's field loop (`admin/index.php`) but the Event form never renders an input for it and nothing reads it. The key is therefore always absent from the stored `bsf_event` option. Dead entry that can be removed.

## Notes

Pre-existing; `admin/index.php` is unchanged between `master` and `dev`. Not related to the rating hardening in #251.

### PR DESCRIPTION
**Pull request by ShubhamGupta05**
Fixes #253

## Summary

- Every settings form on **Rich Snippets → Dashboard** showed **"Sorry, Your changes are not saved!"** after clicking Update without editing a field. Nothing had failed and the settings were intact.
- Root cause: `update_option()` returns `false` both when the write fails **and** when the stored value already matches the submitted one. The handlers treated every `false` as a failure.
- Adds `bsf_save_option()`, which re-reads the option when `update_option()` returns `false` and separates the two cases it conflates.
- `display_status()` gains a third, neutral notice for the unchanged case. A plain truthy value is still treated as a successful save, so any existing caller keeps working.
- Wired into the nine schema forms and the WooCommerce toggle in `admin/index.php`, and into the `submit_color` AJAX handler in `index.php`, which had the same problem (**"Error occured. Settings were not saved !"** on re-saving identical colours).

| Situation | Return | Notice |
|---|---|---|
| Value written | `saved` | Success! Your changes were successfully saved! |
| Already held this value | `unchanged` | No changes to save. |
| Option does not hold the value | `false` | Sorry, Your changes are not saved! |

### Why the helper lives in `functions.php`

`admin/index.php` is only `require_once`d inside `register_custom_menu_page()` on the `admin_menu` hook, which does not fire during `admin-ajax.php`. Defining the helper there and calling it from `submit_color` would be a fatal error. `functions.php` is required unconditionally.

### Note for reviewers

Failure is `false` rather than a `'failed'` string, because any non-empty string is truthy in PHP and would fall through to the success branch. `display_status()` therefore checks `'unchanged' === $status` before the truthy check.

## Test plan

Verified against a real WordPress install by driving the same functions the handlers call.

- [x] Unchanged save on all nine schema options returns `unchanged` and shows the neutral notice.
- [x] A genuine change returns `saved` and stores the new value.
- [x] Saving the same value immediately afterwards returns `unchanged`.
- [x] A diverging stored value returns `false`, so real failures still show the error notice.
- [x] WooCommerce boolean toggle: unchanged saves in both states report `unchanged`, real flips report `saved`.

**Correction:** an earlier revision of this test plan claimed the WooCommerce boolean was verified. It was not. That check ran `update_option()` and `get_option()` in the *same* request, where the option cache holds the exact PHP value just written and hides the serialization round-trip. Real saves happen in fresh requests where the cache is loaded from the database as `'1'`/`''`, and the toggle still produced the failure notice in both states. Caught in review, fixed in 4c3c0f1, and re-verified through the real admin form as a logged-in administrator:

```
BEFORE                                        AFTER
  ON  (already on)  -> Sorry, ...not saved!     ON  (already on)  -> No changes to save.
  ON  (again)       -> Sorry, ...not saved!     ON  (again)       -> No changes to save.
  OFF (real change) -> Success!                 OFF (real change) -> Success!
  OFF (again)       -> Sorry, ...not saved!     OFF (again)       -> No changes to save.
  ON  (real change) -> Success!                 ON  (real change) -> Success!
```

- [x] Schema form driven end to end through the admin UI: real change reports `saved`, resubmitting the same values reports `unchanged`.
- [x] Stored representation unchanged (`'1'` / `''`), so both consumers gating on `empty( $woo_settings )` read exactly as before. No migration.
- [x] Helper resolves in a non-dashboard request, covering the AJAX path.
- [x] Rating regression suite unaffected: 26/26 over HTTP against `admin-ajax.php`.
- [x] `php -l` clean on `index.php`, `functions.php`, `admin/index.php`.

PHPCS passes in CI. Locally the full run needs PHP 8.2 (Local's bundled binary) rather than the system PHP, and still aborts inside WPCS 2.3.0 on `trim(): Passing null`; scoping to individual sniffs reproduces CI exactly and is clean. `composer phpstan` could not be run at all: it aborts during bootstrap with `Cannot redeclare WP_CLI\Dispatcher\CompositeCommand` from `tests/php/stubs/aiosrs-stubs.php`, reproduced on files this branch does not touch.

## Notes

- No version bump.
- Translation template regenerated with `grunt i18n` for the new string.
- #254 fixes the same class of bug in the rating update handler. Both branches regenerate the `.pot`; whichever merges second will most likely need it regenerated rather than hand-resolved.

### CODE REVIEW

**Comment by sushmak02** [comment]
## Review notes

Root-cause analysis is right, and `display_status()` is correctly ordered — `'unchanged'` is tested before the generic truthy branch, so existing callers passing a raw boolean keep working.

**Supply-chain checklist:** clean. All three commits signed (`verified: true`), actor matches author, no build config touched, no dangerous sinks.

**Fatal-error verdict: no fatal risk.** I checked the load order, since `bsf_save_option()` is defined in `functions.php` but called from two other files: `functions.php` is required at top level of `index.php:460`, while `admin/index.php` only loads inside `register_custom_menu_page()` on `admin_menu`, so the helper always exists before use. (`display_status()` being defined at line 1155 after its callers at 1009–1143 is fine — PHP hoists top-level function declarations.)

### The WooCommerce toggle isn't actually fixed

The PR body lists the WooCommerce toggle as wired up, but it's the one form that passes a **boolean** rather than an array of strings (`admin/index.php:1001-1007`):

```php
$args = null;
if ( isset( $_POST['woocommerce_option'] ) ) { $args = true; } else { $args = false; }
$status = bsf_save_option( 'bsf_woocom_setting', $args );
```

WP round-trips scalars through the options table as strings — `true` is stored and read back as `'1'`, `false` as `''`. So the strict comparison in `bsf_save_option()` can never match on this path:

- Toggle **on** twice → `update_option()` returns false (0 rows affected), `get_option()` returns `'1'`, and `'1' === true` is **false** → helper returns `false` → *"Sorry, Your changes are not saved!"*
- Toggle **off** twice → `get_option()` returns `''`, and `'' === false` is **false** → same error notice.

So nine of ten forms are fixed and this one still shows the exact notice the PR sets out to remove, in both toggle states. The nine schema forms are safe because their `$args` is an array of `sanitize_text_field()` strings, which survives serialization with types intact.

Cleanest fix, and consistent with `aiosrs_usage_optin` twelve lines above which already does this: store `'yes'`/`'no'` instead of booleans and keep `===` strict everywhere. Loose `==` in the helper would also work but is a worse trade on a PHP 7.4 floor.

Related, same root cause: `get_option()` returns `false` as its *default* for a missing option, so `bsf_save_option( $opt, false )` on an option that doesn't exist yet reports `'unchanged'`, conflating "absent" with "already false". The `'yes'`/`'no'` change makes that moot.

One caveat on confidence: the boolean round-trip is well-established WP behaviour but I couldn't execute it here — worth a quick manual check of that toggle, since it's the one thing gating my recommendation.

### Affects both this and #256

Both PRs change `@since 1.7.9` to `@since x.x.x` on the **same line** of the `bsf_is_rateable_post()` docblock, and both regenerate the `.pot`. Whichever merges second will conflict on both files — worth sequencing deliberately rather than finding out at merge time. The `@since x.x.x` change itself is correct per our convention.

### Verdict

Needs the WooCommerce toggle converted before it delivers what the description claims. Everything else here is good.

**Comment by ShubhamGupta05** [comment]
Thanks — the WooCommerce finding is correct, and it was a real defect. Fixed in 4c3c0f1.

## Confirmed, including that my original test plan was wrong

You reasoned it out without being able to run it, and it reproduces exactly as described. Driving the real form as a logged-in admin, before and after:

```
########## BEFORE ##########
  ON  (already on)  -> Sorry, Your changes are not saved!
  ON  (again)       -> Sorry, Your changes are not saved!
  OFF (real change) -> Success!
  OFF (again)       -> Sorry, Your changes are not saved!
  ON  (real change) -> Success!

########## AFTER ##########
  ON  (already on)  -> No changes to save.
  ON  (again)       -> No changes to save.
  OFF (real change) -> Success!
  OFF (again)       -> No changes to save.
  ON  (real change) -> Success!
```

Instrumenting the live runtime shows the round-trip you described:

```
before: '1' (string)   update_option -> false
after:  '1' (string)   strict_match: false   both_scalar: true   string_equal: true
```

My original test plan claimed this path was verified. It was not — I had tested `update_option()` and `get_option()` in the **same request**, where the option cache holds the exact PHP value that was just set and masks the serialization entirely. Real saves happen in fresh requests where the cache is populated from the database as `'1'`/`''`. I have corrected that line in the PR description rather than dropping it.

## I did not take the yes/no approach — here is why

Both consumers gate on `empty()`:

```php
// meta-boxes.php:25 and admin/index.php:46
if ( empty( $woo_settings ) ) {   // '' => excluded,  '1' => included
```

`empty( 'no' )` is `false`, so storing `'yes'`/`'no'` would **invert** the setting — turning WooCommerce schema off would read as on. Making that safe means changing both consumers to explicit comparisons *and* migrating every existing install's `'1'`/`''`, or carrying legacy handling at each read site. That is a large blast radius on a setting that controls which post types receive schema markup, for what is a notice-text bug.

Instead the comparison is fixed in the helper:

```php
if ( is_scalar( $args ) && is_scalar( $stored ) ) {
	return (string) $stored === (string) $args ? 'unchanged' : false;
}
```

Strict `===` still runs first and covers arrays, so the nine schema forms are untouched. The scalar branch is a narrower version of the loose `==` you rightly warned against: it only applies when both sides are scalars, never to arrays. Storage format is unchanged, so no migration and no consumer changes.

If you would still prefer `'yes'`/`'no'`, I would rather do it as its own PR with the consumer changes and a migration than fold it in here.

## The absent-option point

Also correct, and now documented in the docblock. `bsf_save_option( $opt, false )` on a missing option returns `'unchanged'`, because `update_option()` early-returns without writing. The reported outcome is accurate — nothing needed changing and the effective setting is already off — so I left the behaviour and described it rather than adding a special case. Saving `true` on an absent option correctly returns `'saved'`.

## One correction on the merge note

The `.pot` conflict is real. The `functions.php` one is not — I ran the merge:

```
Auto-merging functions.php                          <- clean, 0 conflict markers
CONFLICT (content): languages/...rich-snippets.pot  <- 18 conflict markers
```

Both PRs make the **byte-identical** `@since` change on that line, so git's three-way merge resolves it automatically. The duplication is deliberate, so the outcome holds regardless of merge order or if one PR is dropped. Only the `.pot` needs sequencing, and it should be regenerated rather than hand-resolved:

```
npx grunt i18n && git checkout -- admin/bsf-analytics lib/
```

That second command matters — `grunt addtextdomain` rewrites the bundled libraries' own text domains, which should not ship.

Thanks also for the load-order and `display_status()` ordering checks; both match my own analysis.

**Review by sushmak02 — APPROVED** [review]
